### ISSUE TITLE
Resources can only wait for type erased events

### ISSUE THREAD

**Issue by MrBurmark** [LLNL/camp#59]
Currently you have to create a type erased event in order to have a resource wait_for it. Does it make sense to allow resources to wait_for their own event type to avoid this, maybe even wait_for another resource of their own type as syntactic sugar?

**Comment by trws** [comment]
Yeah, that's a combination of problems, since `wait_for` should really take the event by value in which case it would just get auto-wrapped and we also don't have a specific version.  Maybe we should rip off the bandaid and actually fix this set of issues, along with the getting a non-erased event problem, before they get too much usage?
